### PR TITLE
dtoh: Warn about use of C++ keywords as identifiers during -HC generation

### PR DESCRIPTION
**Pull request by ibuclaw**
I've covered:
- C++ operators
- C++ keywords
- C++11 keywords
- C++20 keywords - currently versioned out until the std is finalized.

Testsuite will fail until either violations are fixed, or the headers are generated with `-wi`.

### CODE REVIEW

**Comment by wilzbach** [comment]
> Testsuite will fail until either violations are fixed, or the headers are generated with -wi.

Shouldn't we use `error` here instead?
I don't see how a user could ignore this warning for the header generation.

**Comment by MoonlightSentinel** [comment]
Could dtoh append a _ and override the mangling instead?

**Comment by ibuclaw** [comment]
> > Testsuite will fail until either violations are fixed, or the headers are generated with -wi.
> 
> Shouldn't we use `error` here instead?
> I don't see how a user could ignore this warning for the header generation.

Because it doesn't affect the compilability of the code in D, unlike say having `extern(C++) string fn()`

**Review by UplinkCoder — APPROVED** [review]
Looks good to me

**Comment by MoonlightSentinel** [comment]
Needs a test but otherwise LGTM.

**Comment by ibuclaw** [comment]
I'm not sure why the CI is passing, did someone change the build script or alter the field names in dmd?

**Comment by MoonlightSentinel** [comment]
`cxx-headers-test` is ony run on CircleCI and  linux64 on the auto-tester s.t. the generated header can be compared tor `frontend.h`, other targets cause minor differences.

They report the aforementioned fields:
```
/home/circleci/dmd/src/dmd/dsymbol.d(207): Warning: enum member inline is a keyword in C++
/home/circleci/dmd/src/dmd/dsymbol.d(207):        The generated C++ header will contain identifiers that are keywords in C++
/home/circleci/dmd/src/dmd/tokens.d(117): Warning: enum member and is a special operator in C++
/home/circleci/dmd/src/dmd/tokens.d(118): Warning: enum member or is a special operator in C++
/home/circleci/dmd/src/dmd/tokens.d(119): Warning: enum member xor is a special operator in C++
/home/circleci/dmd/src/dmd/tokens.d(124): Warning: enum member not is a special operator in C++
/home/circleci/dmd/src/dmd/mtype.d(354): Warning: enum member mutable is a keyword in C++
/home/circleci/dmd/src/dmd/parse.d(9218): Warning: enum member or is a special operator in C++
/home/circleci/dmd/src/dmd/parse.d(9219): Warning: enum member xor is a special operator in C++
/home/circleci/dmd/src/dmd/parse.d(9220): Warning: enum member and is a special operator in C++
/home/circleci/dmd/src/dmd/id.d-mixin-26(340): Warning: variable volatile is a keyword in C++
/home/circleci/dmd/src/dmd/id.d-mixin-26(418): Warning: variable unsigned is a keyword in C++
```

**Comment by MoonlightSentinel** [comment]
Needs a rebase.

You should probably disable warnings as errors for `cxx-headers target until we deal with the problematic identifiers:

```patch
diff --git a/src/build.d b/src/build.d
index 47b20c91c..05d34b8f5 100755
--- a/src/build.d
+++ b/src/build.d
@@ -491,7 +491,7 @@ alias buildFrontendHeaders = makeRule!((builder, rule) {
         .deps([dmdDefault])
         .target(env["G"].buildPath("frontend.h"))
         .command([dmdExeFile] ~ flags["DFLAGS"] ~
-            ["-J" ~ env["RES"], "-c", "-o-", "-HCf="~rule.target] ~ dmdSources);
+            ["-J" ~ env["RES"], "-c", "-o-", "-wi", "-HCf="~rule.target] ~ dmdSources);
 });
```

**Comment by MoonlightSentinel** [comment]
Rebased it and used `writeIdentifier` several places that were still using `ident.toChars` directly (see the second commit).
Also added a test case

**Review by RazvanN7 — APPROVED** [review]

**Comment by RazvanN7** [comment]
@MoonlightSentinel Can you rebase again so that we can get this in?

**Comment by ibuclaw** [comment]
@MoonlightSentinel whoops, thanks for picking this up... I've been busy. :-(

**Review by thewilsonator — APPROVED** [review]

File: src/dmd/dtoh.d [hunk 1]
```diff
@@ -387,6 +389,106 @@ public:
...
+                case "consteval":
+                case "constinit":
+                // Concepts-related keywords
+                case "concept":
+                case "requires":
+                // Coroutines-related keywords
+                case "co_await":
+                case "co_yield":
+                case "co_return":
+                    version (none)
```
Comment (ibuclaw) [inline src/dmd/dtoh.d:481]: Since I raised this, we now have C++20 std in the enum.

Comment (MoonlightSentinel) [reply src/dmd/dtoh.d:481]: See #12118

**Pull request merged**